### PR TITLE
Merge all changes for DT version 1.8.2

### PR DESCRIPTION
**Pull request by sakash279**
Changes in this PR include:
- Update Document DB SDK version to Microsoft.Azure.DocumentDB.2.2.1
- Update Tables SDK version to Microsoft.Azure.CosmosDB.Table.2.0.0
- Update Mongo driver version to MongoDB.Driver.2.7.0
- Fix copying of internal fields in azure tables
- Add retry capabilities to azure tables and cosmos DB tables API
- Support new logger : CosmosDB tables logger for error logging
- Error handling improvements for command line parsing
- Add EULA and ThirdParty license file to the codebase

### CODE REVIEW

File: Wpf/Microsoft.DataTransfer.WpfHost/Model/Statistics/ObservableErrorsTransferStatisticsFactory.cs [hunk 1]
```diff
@@ -23,5 +24,10 @@ public ObservableErrorsTransferStatisticsFactory(ITransferStatisticsFactory defa
                 ? new ObservableErrorsTransferStatistics(defaultStatistics, errorDetailsProvider, SynchronizationContext.Current)
                 : defaultStatistics;
         }
+
```
Comment (haitsongmsft) [inline Wpf/Microsoft.DataTransfer.WpfHost/Model/Statistics/ObservableErrorsTransferStatisticsFactory.cs:27]: documentations for the public fields and functions missing.

File: Wpf/Microsoft.DataTransfer.WpfHost/Steps/InfrastructureSetup/InfrastructureConfiguration.cs [hunk 2]
```diff
@@ -35,6 +36,17 @@ public bool OverwriteErrorLog
             set { SetProperty(ref progressUpdateInterval, value); }
         }
 
+        public bool EnableCosmosTableLog
+        {
+            get { return true; }
+        }
+
+        public string CosmosTableLogConnectionString
```
Comment (haitsongmsft) [inline Wpf/Microsoft.DataTransfer.WpfHost/Steps/InfrastructureSetup/InfrastructureConfiguration.cs:44]: similarly, doc for the public functions and fields here and other places.

**Review by haitsongmsft — CHANGES_REQUESTED** [review]

File: Wpf/Microsoft.DataTransfer.WpfHost/Model/Statistics/ObservableErrorsTransferStatisticsFactory.cs (hunk 1 above)
Comment (sakash279) [reply Wpf/Microsoft.DataTransfer.WpfHost/Model/Statistics/ObservableErrorsTransferStatisticsFactory.cs:27]: actually, these new properties along with the others are described in : Core/Microsoft.DataTransfer.ServiceModel/ConfigurationResources.resx

This is the WPF project. I noticed that there is no previous documentation  of other properties. Adding these params here does not make a difference, as the Azure Table adapter (even before my changes) is not supported in the UI/WPF.

Comment (haitsongmsft) [reply Wpf/Microsoft.DataTransfer.WpfHost/Model/Statistics/ObservableErrorsTransferStatisticsFactory.cs:27]: More than these, for example, AzureTableSourceAdapter, looks like we didn't follow at least doc the "public" functions? I am fine if the resx has the documentation for this file.

File: AzureTable/Microsoft.DataTransfer.AzureTable/Sink/TableAPIBulkSinkAdapter.cs [hunk 3]
```diff
@@ -19,17 +20,18 @@ internal sealed class TableAPIBulkSinkAdapter : IDataSinkAdapter
...
+        private readonly int _throughput;
+        private readonly int _maxLengthInBytesPerBatch;
 
         private CloudTable cloudtable;
         private ConcurrentDictionary<string, TableBatchOperation> dict;
         private InputSizeTracker inputSizeTracker;
         private BatchSizeTracker batchSizeTracker;
+        private readonly TableRequestOptions requestOptions;
 
         public int MaxDegreeOfParallelism
```
Comment (haitsongmsft) [inline AzureTable/Microsoft.DataTransfer.AzureTable/Sink/TableAPIBulkSinkAdapter.cs:36]: documentation would be easier to describe the usage of the class and the fields of the public facing members.

**Review by haitsongmsft — APPROVED** [review]

File: Wpf/Microsoft.DataTransfer.WpfHost/Model/Statistics/ObservableErrorsTransferStatisticsFactory.cs (hunk 1 above)
Comment (sakash279) [reply Wpf/Microsoft.DataTransfer.WpfHost/Model/Statistics/ObservableErrorsTransferStatisticsFactory.cs:27]: resolved after discussing, that the current change follows the project documentation pattern.

**Pull request merged**